zookeeper-server/src/main/java/org/apache/zookeeper/ZKWatchManager.java

```diff
@@ -442,10 +442,13 @@ public Set<Watcher> materialize(
         return result;
     }
 
-    private void addPersistentWatches(String clientPath, Set<Watcher> result) {
+    private void addPersistentWatches(String clientPath, Watcher.Event.EventType type, Set<Watcher> result) {
         synchronized (persistentWatches) {
             addTo(persistentWatches.get(clientPath), result);
         }
+        if (type == Watcher.Event.EventType.NodeChildrenChanged) {
```

Comment:
A comment about reason to filter `NodeChildrenChanged` event helps - it's not quite straightforward.

Reply:
Add comments to explain why this is necessary and why this must be done in client side.